File: test/files/tower-of-hanoi.java
// Java with Elan 2.0.0-beta2

public class Global {

static final Int nDiscs = 7; // constant

static final Int delay_ms = 300; // constant

static void main() {
  var stacks = create3Stacks(nDiscs);
  display(stacks); // procedure call
  while (stacks[2].length() != nDiscs) {
    if ((stacks[0].length() % 2) == 0) {
      moveBetween(stacks, 0, 1); // procedure call
      moveBetween(stacks, 0, 2); // procedure call
      moveBetween(stacks, 1, 2); // procedure call
    } else {
      moveBetween(stacks, 0, 2); // procedure call
      moveBetween(stacks, 0, 1); // procedure call
      moveBetween(stacks, 1, 2); // procedure call
    } // end if
  } // end while
} // end main

static void moveBetween(List<List<int>> stacks, int fromStack, int toStack) { // procedure
  var a = stacks[fromStack];
  var b = stacks[toStack];
  if (b.length() < nDiscs) {
    if ((a.length() > 0) && ((b.length() == 0) || (top(a) < top(b)))) {
      var disc = top(a);
      a.removeFirst(disc); // procedure call
      b.append(disc); // procedure call
    } else if (b.length() > 0) {
      var disc = top(b);
      b.removeFirst(disc); // procedure call
      a.append(disc); // procedure call
    } // end if
  } // end if
  display(stacks); // procedure call
} // end procedure

static void display(List<List<int>> stacks) { // procedure
  clearAllDisplays(); // procedure call
  var vg = createVectorGraphics();
  foreach (var i in range(0, 3)) {
    drawStack(stacks[i], i + 1, vg); // procedure call
  } // end foreach
  displayVectorGraphics(vg); // procedure call
  sleep_ms(delay_ms); // procedure call
} // end procedure

static void drawStack(List<int> s, int peg, List<VectorGraphic> vg) { // procedure
  foreach (var n in range(0, s.length())) {
    var discVG = createDisc(s[n], peg, n);
    vg.append(discVG); // procedure call
  } // end foreach
} // end procedure

static RectangleVG createDisc(int disc, int peg, int vertical) { // function
  var r = new RectangleVG();
  return r.withFillColour(colour(disc)).withHeight(3).withWidth(disc*2 + 2).withX((peg - 1)*30 + 20 - disc).withY(50 - vertical*3).withStrokeWidth(0.25);
} // end function

class Test_createDisc {
@Test static void test_createDisc() {
  // Normal cases
  var d = createDisc(5, 2, 4);
  assertEquals(green, d.fillColour);
  assertEquals(0.25, d.strokeWidth);
  assertEquals(3, d.height);
  assertEquals(12, d.width);
  assertEquals(45, d.x);
  assertEquals(38, d.y);
  // Edge cases
  var d2 = createDisc(1, 1, 0);
  assertEquals(red, d2.fillColour);
  assertEquals(0.25, d2.strokeWidth);
  assertEquals(3, d2.height);
  assertEquals(4, d2.width);
  assertEquals(19, d2.x);
  assertEquals(50, d2.y);
  // Error cases - none identified
}} // end test

static List<List<int>> create3Stacks(int nDiscs) { // function
  var s0 = rangeInSteps(nDiscs, 0, -1);
  var s1 = new List<int>();
  var s2 = new List<int>();
  return list(s0, s1, s2);
} // end function

class Test_create3Stacks {
@Test static void test_create3Stacks() {
  var emptyStack = new List<int>();
  // Normal case(s)
  assertEquals(list(list(7, 6, 5, 4, 3, 2, 1), emptyStack, emptyStack), create3Stacks(7));
  // Edge case(s)
  assertEquals(list(list(1), emptyStack, emptyStack), create3Stacks(1));
  assertEquals(list(emptyStack, emptyStack, emptyStack), create3Stacks(0));
  // Error case(s)
  assertEquals("Loop will not terminate when start < end start with negative step", create3Stacks(-1));
}} // end test

static int colour(int disc) { // function
  var colours = list(red, yellow, blue, brown, green, 0xFF9900, 0x6600FF, 0x00CC00, 0x3399FF, 0xFF99CC);
  return colours[disc - 1];
} // end function

class Test_colour {
@Test static void test_colour() {
  // Normal cases
  assertEquals(green, colour(5));
  // Edge cases
  assertEquals(red, colour(1));
  assertEquals(0xFF99CC, colour(10));
  // Error cases
  assertEquals("Out of range index: 10 size: 10", colour(11));
}} // end test

static int top(List<int> s) { // function
  return s[s.length() - 1];
} // end function

class Test_top {
@Test static void test_top() {
  // Normal cases
  assertEquals(1, top(list(3, 2, 1)));
  // Edge cases
  assertEquals(7, top(list(7)));
  // Error cases
  assertEquals("Out of range index: -1 size: 0", top(new List<int>()));
}} // end test
} // end Global

File: test/files/tower-of-hanoi.vb
' VB.NET with Elan 2.0.0-beta-pre1

Const nDiscs = 7

Const delay_ms = 300

Sub main()
  Dim stacks = create3Stacks(nDiscs) ' variable definition
  display(stacks) ' call procedure
  While stacks(2).length() <> nDiscs
    If (stacks(0).length() Mod 2) = 0 Then
      moveBetween(stacks, 0, 1) ' call procedure
      moveBetween(stacks, 0, 2) ' call procedure
      moveBetween(stacks, 1, 2) ' call procedure
    Else
      moveBetween(stacks, 0, 2) ' call procedure
      moveBetween(stacks, 0, 1) ' call procedure
      moveBetween(stacks, 1, 2) ' call procedure
    End If
  End While
End Sub

Sub moveBetween(stacks As List(Of List(Of Integer)), fromStack As Integer, toStack As Integer) ' procedure
  Dim a = stacks(fromStack) ' variable definition
  Dim b = stacks(toStack) ' variable definition
  If b.length() < nDiscs Then
    If (a.length() > 0) And ((b.length() = 0) Or (top(a) < top(b))) Then
      Dim disc = top(a) ' variable definition
      a.removeFirst(disc) ' call procedure
      b.append(disc) ' call procedure
    ElseIf b.length() > 0 Then
      Dim disc = top(b) ' variable definition
      b.removeFirst(disc) ' call procedure
      a.append(disc) ' call procedure
    End If
  End If
  display(stacks) ' call procedure
End Sub

Sub display(stacks As List(Of List(Of Integer))) ' procedure
  clearAllDisplays() ' call procedure
  Dim vg = createVectorGraphics() ' variable definition
  drawStack(stacks(0), 1, vg) ' call procedure
  drawStack(stacks(1), 2, vg) ' call procedure
  drawStack(stacks(2), 3, vg) ' call procedure
  displayVectorGraphics(vg) ' call procedure
  sleep_ms(delay_ms) ' call procedure
End Sub

Sub drawStack(s As List(Of Integer), peg As Integer, vg As List(Of VectorGraphic)) ' procedure
  For Each n In range(0, s.length())
    Dim discVG = createDisc(s(n), peg, n) ' variable definition
    vg.append(discVG) ' call procedure
  Next n
End Sub

Function createDisc(disc As Integer, peg As Integer, vertical As Integer) As RectangleVG
  Dim r = New RectangleVG() ' variable definition
  Return r.withFillColour(colour(disc)).withHeight(3).withWidth(disc*2 + 2).withX((peg - 1)*30 + 20 - disc).withY(50 - vertical*3).withStrokeWidth(0.25)
End Function

<TestMethod> Sub test_createDisc()
  ' Normal cases
  Dim d = createDisc(5, 2, 4) ' variable definition
  Assert.AreEqual(green, d.fillColour)
  Assert.AreEqual(0.25, d.strokeWidth)
  Assert.AreEqual(3, d.height)
  Assert.AreEqual(12, d.width)
  Assert.AreEqual(45, d.x)
  Assert.AreEqual(38, d.y)
  ' Edge cases
  Dim d2 = createDisc(1, 1, 0) ' variable definition
  Assert.AreEqual(red, d2.fillColour)
  Assert.AreEqual(0.25, d2.strokeWidth)
  Assert.AreEqual(3, d2.height)
  Assert.AreEqual(4, d2.width)
  Assert.AreEqual(19, d2.x)
  Assert.AreEqual(50, d2.y)
  ' Error cases - none identified
End Sub

Function create3Stacks(nDiscs As Integer) As List(Of List(Of Integer))
  Dim s0 = rangeInSteps(nDiscs, 0, -1) ' variable definition
  Dim s1 = New List(Of Integer)() ' variable definition
  Dim s2 = New List(Of Integer)() ' variable definition
  Return {s0, s1, s2}
End Function

<TestMethod> Sub test_create3Stacks()
  Dim emptyStack = New List(Of Integer)() ' variable definition
  ' Normal case(s)
  Assert.AreEqual({{7, 6, 5, 4, 3, 2, 1}, emptyStack, emptyStack}, create3Stacks(7))
  ' Edge case(s)
  Assert.AreEqual({{1}, emptyStack, emptyStack}, create3Stacks(1))
  Assert.AreEqual({emptyStack, emptyStack, emptyStack}, create3Stacks(0))
  ' Error case(s)
  Assert.AreEqual("Loop will not terminate when start < end start with negative step", create3Stacks(-1))
End Sub

Function colour(disc As Integer) As Integer
  Dim colours = {red, yellow, blue, brown, green, &HFF9900, &H6600FF, &H00CC00, &H3399FF, &HFF99CC} ' variable definition
  Return colours(disc - 1)
End Function

<TestMethod> Sub test_colour()
  ' Normal cases
  Assert.AreEqual(green, colour(5))
  ' Edge cases
  Assert.AreEqual(red, colour(1))
  Assert.AreEqual(&HFF99CC, colour(10))
  ' Error cases
  Assert.AreEqual("Out of range index: 10 size: 10", colour(11))
End Sub

Function top(s As List(Of Integer)) As Integer
  Return s(s.length() - 1)
End Function

<TestMethod> Sub test_top()
  ' Normal cases
  Assert.AreEqual(1, top({3, 2, 1}))
  ' Edge cases
  Assert.AreEqual(7, top({7}))
  ' Error cases
  Assert.AreEqual("Out of range index: -1 size: 0", top(New List(Of Integer)()))
End Sub
